restli-server/src/main/java/com/linkedin/restli/internal/server/response/ResponseUtils.java

```diff
@@ -84,6 +90,51 @@ static Object translateCanonicalKeyToAlternativeKeyIfNeeded(Object canonicalKey,
     }
   }
 
+  /**
+   * This function will look at the dataSchema to see missing fields in the dataMap.
+   * If the missing field is provided with default values, this function will populate the dataMap
+   * with the key and the default value. If the missing field is not primitive (meaning itself is a
+   * RecordDataSchema) then this function will recursively to fill default values of that field as a record.
+   * This feature is controlled by:
+   * Client: add this query parameter in the URL: $sendDefaults=true
+   * Server: put true to RestLiConfig::isResponseSendDefaultValues
+   * The feature is controlled
+   * @param dataSchema a RecordDataSchema that matches the dataMap
+   * @param dataMap the actual data of the argument dataSchema
+   */
+  public static void fillInDefaultValues(DataSchema dataSchema, DataMap dataMap)
+  {
+    if (dataSchema instanceof TyperefDataSchema)
+    {
+      TyperefDataSchema typerefDataSchema = (TyperefDataSchema) dataSchema;
+      fillInDefaultValues(typerefDataSchema.getDereferencedDataSchema(), dataMap);
+    }
+    else if (dataSchema instanceof RecordDataSchema)
+    {
+      RecordDataSchema recordDataSchema = (RecordDataSchema) dataSchema;
+      for (RecordDataSchema.Field field : recordDataSchema.getFields())
+      {
+        if (!dataMap.containsKey(field.getName()))
+        {
+          if (field.getDefault() != null)
+          {
+            dataMap.put(field.getName(), field.getDefault());
+          }
+          else
+          {
+            DataSchema fieldSchema = field.getType();
+            if (fieldSchema instanceof RecordDataSchema)
+            {
+              DataMap fieldDataMap = new DataMap();
+              fillInDefaultValues(fieldSchema, fieldDataMap);
+              dataMap.put(field.getName(), fieldDataMap);
+            }
+          }
```

Comment:
This is not needed. You need to traverse the nested schemas only if the value is present at the top-level.

Consider this:
record Foo {
  f1: int = 5
  f2: optional int
 }

record Bar {
   b1: Foo = { f2: 10 }
   b2: optional Foo
   b3: Foo
}

So the scenarios are (top-level is Bar):
1. 
{
  b1: {
    f1: 1
  },
  b2: {}
}
Output with defaults filled in should be:
{
  b1: {
    f1: 1
  },
   b2: {
    f1: 5,     // Filled using default value in Foo
  }
}
b3 is not present in the input data and so shouldn't be filled in.

2.
{
  b3: {
    f2: 1
  }
}
Output with defaults filled in should be:
{
 b1: {.         // Filled because of default value in Bar
   f1: 5,        // Filled using default value in Foo
    f2: 10     // Filled using default value in Bar
  },
  b3: {
    f1: 5,     // Filled using default value in Foo
    f2: 1
  }
}

Basically, fill in the defaults for fields only if the record is present or record is filled in using a default. Do not create a record only because a nested field specifies default value.

Reply:
I think this is addressed, had put exact same cases in test